File: Leetcode/PathSum.py
# Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.

class Solution(object):
    def hasPathSum(self, root, targetSum):
        def dfs(node, curSum):
            if not node:
                return False
            
            curSum += node.val
            if not node.left and not node.right:
                return curSum == targetSum
            
            return dfs(node.left, curSum) or dfs(node.right, curSum)
        return dfs(root, 0)

File: Leetcode/PathSum.cs
// Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.

public class Solution 
{
    public bool HasPathSum(TreeNode root, int targetSum) 
    {
        return DFS(root, 0, targetSum);
    }
    
    public static bool DFS(TreeNode node, int curSum, int targetSum)
    {
        if (node == null)
            return false;
        
        curSum += node.val;
        if (node.left == null && node.right == null)
            return curSum == targetSum;
        
        return DFS(node.left, curSum, targetSum) || DFS(node.right, curSum, targetSum);
    }
}